### DISCUSSION TITLE
How can I explicitly set intermediate filename prefix to avoid confliction?

### DISCUSSION THREAD

**Discussion by xq114** [xmake-io/xmake#1569] [Q&A] [answered]
现有若干c文件需要在不同的define下编译并链接：
```
target("test")
    add_files("src/*.c", {defines="DINT"})
    add_files("src/*.c", {defines="DLONG"})
```
这些文件的中间文件名都是%.c.o，导致两种定义的输出被覆盖，实际链接时并没有把两种情况编译的文件都链接进去。如何设置中间文件的前缀/后缀，避免这一冲突？

**Comment by waruqi** [comment, marked as answer]
没懂，你两个 add_files 匹配的同一批文件，当然冲突了，不想冲突，自己改下文件名或者路径

**Reply by xq114** [reply]
就是同一批文件，用不同的define生成不同的object file，再把所有的object file链接

**Reply by xq114** [reply]
这种用法是用来在C里面模拟C++的模板的，见https://github.com/DrTimothyAldenDavis/SuiteSparse/blob/master/CHOLMOD/Include/cholmod_internal.h#L172-L183

**Reply by waruqi** [reply]
同一文件只能唯一对应一个 obj path，只能自己写 rule，额外 copy 一份出来，用不同的名字添加到 add_files 里面去。。或者直接在 rule 走编译，同时编译一份，添加到 target:objectfiles

**Reply by xq114** [reply]
怎样把defines作为参数传给rule？

**Reply by waruqi** [reply]
> 怎样把defines作为参数传给rule？

参考下 protobuf rule https://github.com/xmake-io/xmake/blob/master/tests/projects/c%2B%2B/protobuf/xmake.lua
